evm/p2p/lightchain.py

```diff
@@ -0,0 +1,76 @@
+import asyncio
+import logging
+
+from async_lru import alru_cache
+
+from eth_utils import encode_hex
+
+from evm.chains import Chain
+from evm.exceptions import (
+    BlockNotFound,
+)
+from evm.p2p.utils import gen_request_id
+
+
+class OnDemandDataBackend:
+    logger = logging.getLogger("evm.p2p.lightchain.OnDemandDataBackend")
+    # TODO:
+    # 1. Use the PeerManager to maintain a set of connected peers
+    # 2. Implement a queue of requests and a distributor which picks items from that queue and
+    # sends them to one of our peers, ensuring the selected peer has the info we want and
+    # respecting the flow control rules
+
+    @alru_cache(maxsize=1024)
+    @asyncio.coroutine
+    def get_block_by_hash(self, block_hash):
+        peer = yield from self.get_peer()
+        self.logger.debug("Fetching block {} from peer {}".format(encode_hex(block_hash), peer))
+        request_id = gen_request_id()
+        peer.les_proto.send_get_block_bodies([block_hash], request_id)
+        reply = yield from peer.wait_for_reply(request_id)
+        if len(reply['bodies']) == 0:
+            raise BlockNotFound("No block with hash {} found".format(block_hash))
+        return reply['bodies'][0]
+
+    @asyncio.coroutine
+    def get_peer(self):
+        raise NotImplementedError("TODO")
+
+    @asyncio.coroutine
+    def stop(self):
+        raise NotImplementedError("TODO")
+
+
+class LightChain(Chain):
+    on_demand_data_backend_class = None
```

Comment:
Is this intended to be set to `None` or should it be the `OnDemandDataBackend`?

Reply:
I think eventually we'll provide a default, yeah, but since OnDemandDataBackend is currently incomplete, I thought it'd be better to set it to None and force us to specify a working one (which at this time will be hack-ish like the one used for the RPC server) via `.configure()`